cfbs/core.py

```diff
@@ -375,60 +371,95 @@ def _add_modules(
         # Should have a dict with keys for everything in to_add:
         assert not any((k not in added_by for k in to_add))
 
-        # Print error and exit if there are unknown modules:
-        missing = [m for m in to_add if not m.startswith("./") and m not in index]
-        if missing:
-            user_error("Module(s) could not be found: %s" % ", ".join(missing))
+        return added_by
 
-        # If some modules were added as deps previously, mark them as user requested:
+    def _update_added_by(self, requested, added_by):
         for module in self["build"]:
-            if module["name"] in to_add:
+            if module["name"] in requested:
                 new_added_by = added_by[module["name"]]
                 if new_added_by == "cfbs add":
                     module["added_by"] = "cfbs add"
 
-        # Filter modules which are already added:
+    def _filter_modules_to_add(self, modules):
         added = [m["name"] for m in self["build"]]
         filtered = []
-        for module in to_add:
-            user_requested = added_by[module] == "cfbs add"
-            if module in [*added, *filtered]:
-                if user_requested:
-                    print("Skipping already added module: %s" % module)
-                continue
-            filtered.append(module)
+        for module in modules:
+            assert module not in filtered
+            if module in added:
+                print("Skipping already added module: %s" % module)
+            else:
+                filtered.append(module)
+        return filtered
+
+    def _find_dependencies(self, modules, exclude):
+        assert type(modules) is list
+        assert type(exclude) is list
+        exclude = modules + exclude
+        exclude_names = [m["name"] for m in exclude]
+        index = self.index
+        dependencies = []
+        for module in modules:
+            if "dependencies" in module:
+                for dep in module["dependencies"]:
+                    if dep not in exclude_names:
+                        m = index.get_module_object(dep, module["name"])
+                        dependencies.append(m)
+                        exclude_names.append(dep)
```

Comment:
Might be nicer and more "pythonic" to check the opposite conditions and do `continue` to reduce nesting here.

Reply:
Yep, thanks, reduced nesting here: https://github.com/cfengine/cfbs/pull/93/commits/c0d37d84e08518893d4950b3264f58a83e542ab0